// @mui/x-charts/ScatterChart/async/ScatterAsync.mjs
'use client';

import * as React from 'react';
import PropTypes from 'prop-types';
import { ScatterAsyncBatch } from "./ScatterAsyncBatch.mjs";
import { useStore } from "../../internals/store/useStore.mjs";
import { selectorProgressiveBatchSize, selectorProgressiveSeriesRevealedBatches } from "../../internals/plugins/featurePlugins/useProgressiveRendering/index.mjs";
import { selectorScatterSeriesRenderData } from "./scatterRenderData.selectors.mjs";

/**
 * @ignore - internal component.
 */
import { jsx as _jsx } from "react/jsx-runtime";
function ScatterAsync(props) {
  const {
    series,
    colorGetter,
    onItemClick,
    slots,
    slotProps,
    classes
  } = props;
  const store = useStore();
  const batchSize = store.use(selectorProgressiveBatchSize);
  const revealedBatches = store.use(selectorProgressiveSeriesRevealedBatches, series.id);
  // Size batches by the number of *visible* points so that zooming in (which
  // shrinks the filtered set in the selector) collapses the progressive wave
  // into a single tick once everything fits in one batch.
  const renderData = store.use(selectorScatterSeriesRenderData, series.id);
  const count = renderData?.count ?? 0;
  const nBatches = count === 0 ? 0 : Math.ceil(count / Math.max(1, batchSize));
  const batches = [];
  for (let b = 0; b < nBatches; b += 1) {
    const start = b * batchSize;
    const end = Math.min(count, start + batchSize);
    batches.push(/*#__PURE__*/_jsx(ScatterAsyncBatch, {
      series: series,
      colorGetter: colorGetter,
      onItemClick: onItemClick,
      slots: slots,
      slotProps: slotProps,
      start: start,
      end: end,
      classes: classes,
      revealed: b < revealedBatches
    }, b));
  }
  return /*#__PURE__*/_jsx(React.Fragment, {
    children: batches
  });
}
process.env.NODE_ENV !== "production" ? ScatterAsync.propTypes = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  classes: PropTypes.object,
  color: PropTypes.string.isRequired,
  colorGetter: PropTypes.func.isRequired,
  onItemClick: PropTypes.func,
  series: PropTypes.object.isRequired,
  slotProps: PropTypes.object,
  slots: PropTypes.object,
  xScale: PropTypes.func.isRequired,
  yScale: PropTypes.func.isRequired
} : void 0;
export { ScatterAsync };
