// @mui/x-charts/ScatterChart/async/ScatterAsync.js
"use strict";
'use client';

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScatterAsync = ScatterAsync;
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _ScatterAsyncBatch = require("./ScatterAsyncBatch");
var _useStore = require("../../internals/store/useStore");
var _useProgressiveRendering = require("../../internals/plugins/featurePlugins/useProgressiveRendering");
var _scatterRenderData = require("./scatterRenderData.selectors");
var _jsxRuntime = require("react/jsx-runtime");
/**
 * @ignore - internal component.
 */function ScatterAsync(props) {
  const {
    series,
    colorGetter,
    onItemClick,
    slots,
    slotProps,
    classes
  } = props;
  const store = (0, _useStore.useStore)();
  const batchSize = store.use(_useProgressiveRendering.selectorProgressiveBatchSize);
  const revealedBatches = store.use(_useProgressiveRendering.selectorProgressiveSeriesRevealedBatches, series.id);
  // Size batches by the number of *visible* points so that zooming in (which
  // shrinks the filtered set in the selector) collapses the progressive wave
  // into a single tick once everything fits in one batch.
  const renderData = store.use(_scatterRenderData.selectorScatterSeriesRenderData, series.id);
  const count = renderData?.count ?? 0;
  const nBatches = count === 0 ? 0 : Math.ceil(count / Math.max(1, batchSize));
  const batches = [];
  for (let b = 0; b < nBatches; b += 1) {
    const start = b * batchSize;
    const end = Math.min(count, start + batchSize);
    batches.push(/*#__PURE__*/(0, _jsxRuntime.jsx)(_ScatterAsyncBatch.ScatterAsyncBatch, {
      series: series,
      colorGetter: colorGetter,
      onItemClick: onItemClick,
      slots: slots,
      slotProps: slotProps,
      start: start,
      end: end,
      classes: classes,
      revealed: b < revealedBatches
    }, b));
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(React.Fragment, {
    children: batches
  });
}
process.env.NODE_ENV !== "production" ? ScatterAsync.propTypes = {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // | To update them edit the TypeScript types and run "pnpm proptypes"  |
  // ----------------------------------------------------------------------
  classes: _propTypes.default.object,
  color: _propTypes.default.string.isRequired,
  colorGetter: _propTypes.default.func.isRequired,
  onItemClick: _propTypes.default.func,
  series: _propTypes.default.object.isRequired,
  slotProps: _propTypes.default.object,
  slots: _propTypes.default.object,
  xScale: _propTypes.default.func.isRequired,
  yScale: _propTypes.default.func.isRequired
} : void 0;
