// @mui/x-charts/internals/plugins/featurePlugins/useProgressiveRendering/useProgressiveRendering.selectors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sameSeriesIds = sameSeriesIds;
exports.selectorShouldUseProgressiveRenderer = exports.selectorProgressiveTotalRounds = exports.selectorProgressiveSeriesRevealedBatches = exports.selectorProgressiveRevealedRounds = exports.selectorProgressivePlans = exports.selectorProgressiveBatchSize = exports.selectorProgressiveAggregate = void 0;
var _store = require("@mui/x-internals/store");
var _useChartSeries = require("../../corePlugins/useChartSeries");
var _useChartExperimentalFeature = require("../../corePlugins/useChartExperimentalFeature");
/**
 * Total point count above which the auto (`renderer` unset) mode switches to
 * the progressive renderer. Below it the synchronous renderer is used, since
 * the progressive paint's overhead is not worth it for small datasets.
 */
const PROGRESSIVE_POINT_THRESHOLD = 20000;

/**
 * Target number of reveal commits. Each commit repaints every already-painted
 * circle, so the total progressive wall time is roughly `(C + 1) / 2` times a
 * single synchronous render (where `C` is the number of commits). Targeting
 * `5` commits keeps the progressive paint at roughly 2–3× the sync render time.
 */
const TARGET_PROGRESSIVE_COMMITS = 5;

/**
 * Lower bound for the per-tick reveal budget (total points across every
 * series). Prevents tiny commits whose React overhead would dominate.
 */
const MIN_BATCH_TOTAL = 1000;

/**
 * Upper bound for the per-tick reveal budget (total points across every
 * series). Prevents a single commit from blocking the main thread for too
 * long; very large datasets simply use more commits.
 */
const MAX_BATCH_TOTAL = 10000;

/**
 * Per-series points revealed per tick, derived from the total point count
 * across visible series and the number of visible series. The total per-tick
 * budget aims for {@link TARGET_PROGRESSIVE_COMMITS} commits, clamped by
 * {@link MIN_BATCH_TOTAL} / {@link MAX_BATCH_TOTAL}, then split evenly across
 * the visible series so every series progresses together.
 */
const getEffectiveBatchSize = (nSeries, totalPoints) => {
  const safeSeries = Math.max(1, nSeries);
  const safePoints = Math.max(1, totalPoints);
  const totalPerTick = Math.min(MAX_BATCH_TOTAL, Math.max(MIN_BATCH_TOTAL, Math.ceil(safePoints / TARGET_PROGRESSIVE_COMMITS)));
  return Math.max(1, Math.floor(totalPerTick / safeSeries));
};
const selectorProgressiveState = state => state.progressiveRendering;

/** Map of registered plots → their set of series ids. */
const selectorProgressivePlans = exports.selectorProgressivePlans = (0, _store.createSelector)(selectorProgressiveState, s => s?.plans);

/** Total number of rounds revealed across all plots so far. */
const selectorProgressiveRevealedRounds = exports.selectorProgressiveRevealedRounds = (0, _store.createSelector)(selectorProgressiveState, s => s?.revealedRounds ?? 0);

/** Point count of a series, looked up across every processed series type. */
function getSeriesPointCount(processedSeries, seriesId) {
  for (const type in processedSeries) {
    if (!Object.hasOwn(processedSeries, type)) {
      continue;
    }
    const item = processedSeries[type]?.series[seriesId];
    if (item) {
      return Array.isArray(item.data) ? item.data.length : 0;
    }
  }
  return 0;
}

/**
 * Aggregated view of every registered plan: the per-series batch counts, the
 * total number of rounds, and the per-series batch size (so every consumer
 * sizes its batches the same way). Point counts are read straight from the
 * processed series rather than carried by the registration.
 */
const selectorProgressiveAggregate = exports.selectorProgressiveAggregate = (0, _store.createSelectorMemoized)(selectorProgressivePlans, _useChartSeries.selectorChartSeriesProcessed, function selectorProgressiveAggregate(plans, processedSeries) {
  const nBatchesBySeries = new Map();
  if (!plans || plans.size === 0) {
    return {
      nBatchesBySeries,
      totalRounds: 0,
      batchSize: MIN_BATCH_TOTAL
    };
  }
  let nSeries = 0;
  let totalPoints = 0;
  const pointCounts = new Map();
  plans.forEach(seriesIds => {
    seriesIds.forEach(seriesId => {
      const nPoints = getSeriesPointCount(processedSeries, seriesId);
      pointCounts.set(seriesId, nPoints);
      nSeries += 1;
      totalPoints += nPoints;
    });
  });
  const batchSize = getEffectiveBatchSize(nSeries, totalPoints);
  let totalRounds = 0;
  pointCounts.forEach((nPoints, seriesId) => {
    const n = Math.max(1, Math.ceil(nPoints / batchSize));
    nBatchesBySeries.set(seriesId, n);
    if (n > totalRounds) {
      totalRounds = n;
    }
  });
  return {
    nBatchesBySeries,
    totalRounds,
    batchSize
  };
});
const selectorProgressiveBatchSize = exports.selectorProgressiveBatchSize = (0, _store.createSelector)(selectorProgressiveAggregate, a => a.batchSize);
const selectorProgressiveTotalRounds = exports.selectorProgressiveTotalRounds = (0, _store.createSelector)(selectorProgressiveAggregate, a => a.totalRounds);

/**
 * How many of `seriesId`'s own batches are revealed so far. Capped at that
 * series' total batch count, so series with fewer batches simply stop
 * progressing while longer ones keep filling in.
 */
const selectorProgressiveSeriesRevealedBatches = exports.selectorProgressiveSeriesRevealedBatches = (0, _store.createSelector)(selectorProgressiveAggregate, selectorProgressiveRevealedRounds, function selectorProgressiveSeriesRevealedBatches(agg, revealed, seriesId) {
  return Math.min(revealed, agg.nBatchesBySeries.get(seriesId) ?? 0);
});

/**
 * Whether `seriesIds` should be rendered progressively given the requested
 * `renderer`:
 * - `svg-single` / `svg-batch`: never (those are non-progressive renderers).
 * - `svg-progressive`: always.
 * - unset (auto): only when the `progressiveRendering` experimental feature is
 *   enabled and the total point count is above
 *   {@link PROGRESSIVE_POINT_THRESHOLD}. The flag keeps the auto behavior
 *   opt-in so the default (`svg-single`) stays non-breaking.
 */
const selectorProgressiveRenderingEnabled = state => (0, _useChartExperimentalFeature.selectorChartExperimentalFeaturesState)(state, 'progressiveRendering');
const selectorShouldUseProgressiveRenderer = exports.selectorShouldUseProgressiveRenderer = (0, _store.createSelector)(_useChartSeries.selectorChartSeriesProcessed, selectorProgressiveRenderingEnabled, (processedSeries, progressiveRenderingEnabled, seriesIds, renderer) => {
  if (renderer === 'svg-single' || renderer === 'svg-batch') {
    return false;
  }
  if (renderer === 'svg-progressive') {
    return true;
  }
  if (!progressiveRenderingEnabled) {
    return false;
  }
  let totalPoints = 0;
  for (const seriesId of seriesIds) {
    totalPoints += getSeriesPointCount(processedSeries, seriesId);
  }
  return totalPoints > PROGRESSIVE_POINT_THRESHOLD;
});

/** Order-sensitive equality between two registered series-id sets. */
function sameSeriesIds(a, b) {
  if (!a || a.length !== b.length) {
    return false;
  }
  for (let i = 0; i < a.length; i += 1) {
    if (a[i] !== b[i]) {
      return false;
    }
  }
  return true;
}
